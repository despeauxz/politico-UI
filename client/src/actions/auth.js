import instance from "../config/axios";
import errorHandler from '../helpers/errorHandler';
import authAPI from '../utils/api/authAPI';
import {
    AUTHENTICATING,
    AUTHENTICATED,
    IS_AUTHENTICATED,
    UNAUTHENTICATED,
    AUTHENTICATION_ERROR,
    CLEAR_AUTH_ERROR,
    SIGNIN_SUCCESS,
    SIGNIN_ERROR,
    SIGNUP_SUCCESS,
    SIGNUP_ERROR,
    SET_AUTH_WORKING,
    GET_CURRENT_USER,
    UNSET_AUTH_WORKING
} from "../constants/actionTypes";

export const authenticating = () => ({
    type: AUTHENTICATING
});

export const isAuthenticated = () => ({
    type: IS_AUTHENTICATED
});

export const getCurrentUser = user => ({
    type: GET_CURRENT_USER,
    payload: user
});

export const setAuthWorking = () => ({
    type: SET_AUTH_WORKING
});

export const unsetAuthWorking = () => ({
    type: UNSET_AUTH_WORKING
});

export const authenticationSuccess = user => ({
    type: AUTHENTICATED,
    payload: user
});

export const authenticationFailure = error => ({
    type: AUTHENTICATION_ERROR,
    error
});

export const clearAuthError = () => ({
    type: CLEAR_AUTH_ERROR
});

export const signinSuccess = user => ({
    type: SIGNIN_SUCCESS,
    payload: user
});

export const signinFailure = error => ({
    type: SIGNIN_ERROR,
    payload: error
});

export const signupSuccess = user => ({
    type: SIGNUP_SUCCESS,
    payload: user
});

export const signupFailure = error => ({
    type: SIGNUP_ERROR,
    payload: error
});

export const resetUser = () => ({
    type: UNAUTHENTICATED
});

export const auth = (type, user) => async (dispatch) => {
    try {

        dispatch(authenticating());
        const response = await authAPI(type, user);
        localStorage.setItem('jwtToken', response.data.data.token);

        const dispatchType = type === 'signup' ? signupSuccess : signinSuccess;
        dispatch(dispatchType(response.data.data.user));
    } catch (error) {
        const errorResponse = errorHandler(error);
        const dispatchType = type === 'signup' ? signupFailure : signinFailure;

        dispatch(dispatchType(errorResponse.response));
    }
};

export const authenticateUser = () => async (dispatch) => {
    try {
        dispatch(authenticating());
    
        const res = await instance.get('/auth/refresh_token');
    
        localStorage.setItem('jwtToken', res.data.token);
    
        dispatch(authenticationSuccess(res.data.user));
    } catch (error) {
        const errorResponse = errorHandler(error);
    
        localStorage.removeItem('jwtToken');
    
        dispatch(authenticationFailure(errorResponse.response));
    }
};

export const logout = () => (dispatch) => {
    localStorage.removeItem("jwtToken");
    dispatch(resetUser());
};
