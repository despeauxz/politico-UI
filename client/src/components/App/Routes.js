import React from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import Welcome from '@containers/pages/Welcome';
import Auth from '@containers/pages/Auth';
import NotFound from '@components/pages/NotFound';
import Authenticator from '@containers/hoc/Authenticator';
import Parties from '@containers/pages/Parties';
import Office from '@containers/pages/Office';
<<<<<<< HEAD
import Profile from '@containers/pages/Profile';
=======
>>>>>>> develop


const Routes = () => 
    <BrowserRouter>
        <Switch>
            <Route exact path="/" component={Welcome} />
            <Route exact path="/auth/signup" render={props => <Auth {...props} type="signup" />} />
            <Route exact path="/auth/login" render={props => <Auth {...props} type="login" />} />
            <Route exact path="/profile" component={Authenticator(Profile)} />
            <Route exact path="/parties" component={Authenticator(Parties)} />
            <Route exact path="/office" component={Authenticator(Office)} />
            <Route path="/*" component={NotFound} />
        </Switch>
    </BrowserRouter>;

export default Routes;
