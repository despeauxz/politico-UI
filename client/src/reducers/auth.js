import { 
    AUTHENTICATING,
    AUTHENTICATED,
    UNAUTHENTICATED,
    AUTHENTICATION_ERROR,
    CLEAR_AUTH_ERROR,
    SIGNIN_SUCCESS,
    SIGNIN_ERROR,
    SIGNUP_SUCCESS,
    SIGNUP_ERROR,
    SET_AUTH_WORKING,
    UNSET_AUTH_WORKING,
} from "../constants/actionTypes";

const initialState = {
    isAuthenticated: !!localStorage.getItem('jwtToken'),
    error: null,
    user: {},
    loading: false,
    working: false,
};

export default (state = initialState, action) => {
    switch (action.type) {
    case AUTHENTICATING:
        return { 
            ...state,
            loading: true
        };
    case SET_AUTH_WORKING:
        return {
            ...state,
            working: true
        };
    case UNSET_AUTH_WORKING:
        return {
            ...state,
            working: false
        };
    case AUTHENTICATED:
    case SIGNIN_SUCCESS:
    case SIGNUP_SUCCESS:
        return {
            ...state,
            isAuthenticated: true,
            user: action.payload,
            error: null,
            loading: false
        };
    case UNAUTHENTICATED:
    case CLEAR_AUTH_ERROR:
        return {
            ...state,
            isAuthenticated: false,
            error: null,
            loading: false
        };
    case AUTHENTICATION_ERROR:
    case SIGNIN_ERROR:
    case SIGNUP_ERROR:
        return {
            ...state,
            isAuthenticated: false,
            user: {},
            error: action.payload,
            loading: false
        };
    default:
        return state;
    }
};