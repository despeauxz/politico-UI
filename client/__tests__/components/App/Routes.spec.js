import React from 'react';
import Routes from '@components/App/Routes';
import { createStore, combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';
import { BrowserRouter, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import auth from '@reducers/auth';

describe('Routes', () => {
    let store, history;

    beforeEach(() => {
        store = createStore(combineReducers({
            router: routerReducer,
            auth
        }));

        history = createMemoryHistory();
    });

    it('should render welcomepage', () => {
        const comp = 
            <Provider store={store}>
                <BrowserRouter>
                    <Switch>
                        <Routes dispatch={jest.fn()} />
                    </Switch>
                </BrowserRouter>
            </Provider>;

        const wrapper = mount(comp);
  
        expect(wrapper.find('Welcome')).toBeTruthy();
    });

    it('should render login page', () => {
        const comp = 
          <Provider store={store}>
              <BrowserRouter>
                  <Switch>
                      <Routes dispatch={jest.fn()} />
                  </Switch>
              </BrowserRouter>
          </Provider>;
        const wrapper = mount(comp);
        history.push('/auth/login');

        expect(wrapper.find('Auth')).toBeTruthy();
    });

    it('should render signup page', () => {
        const comp = 
        <Provider store={store}>
            <BrowserRouter>
                <Switch>
                    <Routes dispatch={jest.fn()} />
                </Switch>
            </BrowserRouter>
        </Provider>;
        const wrapper = mount(comp);
        history.push('/auth/signup');

        expect(wrapper.find('Auth')).toBeTruthy();
    });

    it('should render parties page', () => {
        const comp = 
        <Provider store={store}>
            <BrowserRouter>
                <Switch>
                    <Routes dispatch={jest.fn()} />
                </Switch>
            </BrowserRouter>
        </Provider>;
        const wrapper = mount(comp);
        history.push('/parties');

        expect(wrapper.find('Parties')).toBeTruthy();
    });

    it('should render not found page', () => {
        const comp = 
        <Provider store={store}>
            <BrowserRouter>
                <Switch>
                    <Routes dispatch={jest.fn()} />
                </Switch>
            </BrowserRouter>
        </Provider>;
        const wrapper = mount(comp);
        history.push('/not-found');

        expect(wrapper.find('NotFound')).toBeTruthy();
    });
  

    it('should render the Routes component correctly', () => {
        const shallowWrapper = shallow(<Routes />);

        expect(toJson(shallowWrapper)).toMatchSnapshot();
    });
});