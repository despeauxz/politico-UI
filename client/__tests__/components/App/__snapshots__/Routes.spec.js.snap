// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes should render the Routes component correctly 1`] = `
<BrowserRouter>
  <Switch>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Welcome)",
          "type": [Function],
        }
      }
      exact={true}
      path="/"
    />
    <Route
      exact={true}
      path="/auth/signup"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/auth/login"
      render={[Function]}
    />
    <Route
      component={[Function]}
      path="/*"
    />
  </Switch>
</BrowserRouter>
`;
